sklearn/decomposition/tests/test_nmf.py

```diff
@@ -933,3 +933,29 @@ def test_NMF_inverse_transform_W_deprecation():
 
     with pytest.warns(FutureWarning, match="Input argument `W` was renamed to `Xt`"):
         est.inverse_transform(W=Xt)
+
+
+def test_NMF_n_components_auto():
+    W_true = np.random.rand(6, 2)
+    H_true = np.random.rand(2, 5)
+    X = np.dot(W_true, H_true)
+
+    # Neither H_true nor W_true are set, n_components is set to n_features
+    W, H, _ = non_negative_factorization(X, n_components="auto")
+    assert W.shape[1] == X.shape[1] and H.shape[0] == X.shape[1]
+
+    # H_true is set, inferring n_components from it
+    W, H, _ = non_negative_factorization(X, H=H_true, n_components="auto")
+    assert H.shape[0] == H_true.shape[0]
+
+    # H_true is set, inferring n_components from it
+    W, H, _ = non_negative_factorization(X, W=W_true, n_components="auto")
+    assert W.shape[1] == W_true.shape[1]
```

Comment:
This is not testing what you expect. Since you didn't set `init="custom"`, W and H are ignored.

Maybe we should raise if ``init!="custom"`` and W or H is passed.

Reply:
I would say that W and H are ignored if you don't set `init="custom"` or you don't set `update_H=False`

What we could do (but I'll let you decide if this is in or out of the scope of this PR) is top do a check here
https://github.com/scikit-learn/scikit-learn/blob/0e1f170d8a392c6c43a67edf0d7f46150956aacf/sklearn/decomposition/_nmf.py#L1213-L1216
and warn the user if `W` or `H` was provided and will not be used (because they are initialised regardless)

Reply:
A warning is good, yes. I'd do that in this PR.

We also need a warning for W only just above, when update_H is False and W is not None.

Reply:
done in https://github.com/scikit-learn/scikit-learn/pull/26634/commits/82fbc4c064481d573199fdd0dc85e76359f382f0